pkg/nsx/services/builder.go

```diff
@@ -639,6 +593,68 @@ func (service *SecurityPolicyService) buildRuleDstGroup(obj *v1alpha1.SecurityPo
 	return &ruleDstGroup, ruleDstGroupPath, err
 }
 
+func (service *SecurityPolicyService) buildEntry(port v1alpha1.SecurityPolicyPort,
+	portIP util.Address) *data.StructValue {
+	var portRange string
+	sourcePorts := data.NewListValue()
+	destinationPorts := data.NewListValue()
+
+	// In case that the destination_port in NSX-T is 0.
+	endPort := port.EndPort
+	if endPort == 0 {
+		portRange = fmt.Sprint(portIP.Port)
+	} else {
+		portRange = fmt.Sprintf("%d-%d", portIP.Port, endPort)
+	}
+	destinationPorts.Add(data.NewStringValue(portRange))
+
+	serviceEntry := data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"source_ports":      sourcePorts,
+			"destination_ports": destinationPorts,
+			"l4_protocol":       data.NewStringValue(string(port.Protocol)),
+			"resource_type":     data.NewStringValue("L4PortSetServiceEntry"),
+			// Adding the following default values to make it easy when compare the
+			// existing object from store and the new built object
+			"marked_for_delete": data.NewBooleanValue(false),
+			"overridden":        data.NewBooleanValue(false),
+		},
+	)
+	log.V(2).Info("built service entry", "serviceEntry", serviceEntry)
+	return serviceEntry
+}
+
+func (service *SecurityPolicyService) buildRuleBasicInfo(obj *v1alpha1.SecurityPolicy,
+	rule *v1alpha1.SecurityPolicyRule, ruleIdx int, portIdx int, dupPortIdx int) (*model.Rule, error) {
+	nsxRuleID := service.buildRuleID(obj, ruleIdx)
+	nsxRuleName := service.buildRuleName(obj, rule, ruleIdx)
+	ruleAction, err := getRuleAction(rule)
+	if err != nil {
+		return nil, err
+	}
+	ruleDirection, err := getRuleDirection(rule)
+	if err != nil {
+		return nil, err
+	}
+	sequence := int64(ruleIdx)
+	rID := fmt.Sprintf("%s_%d_%d", nsxRuleID, portIdx, dupPortIdx)
```

Comment:
Are we able to handle upgrade case here since ruleID is changed with last released version? 
For example, in the last version, we create a ruleA with ruleID := service.buildRuleID(obj, ruleIdx)
If user upgrade the new nsx operator with new named port support, so, users change security policy ruleA by adding named port in. This is a secenario where users add named port to existing rule. 
To my understanding, the nsx operator will create a new rule 
with new id like:
ruleID= mt.Sprintf("%s_%d_%d", nsxRuleID, portIdx, dupPortIdx)

Can NSX will override the older RuleA or create a new RuleAA one and delete RuleA?

Reply:
In our current design, we used use ipSet method to support namedport and non namedport, is it? So, for both namedport and non namedport, the groupname is like: named-port-policy-0-0-0-ipset.
Do we consider to add portname-portnumber into to ruleid/rulename for help user to better understand the rule's group as NCP namedport did?